### PR TITLE
Upgrade `typescript` to 4.9.5

### PR DESCRIPTION
**Pull request by kevin940726**
## What?

Upgrade `typescript` to the latest version (`4.4.2` -> `4.9.5`).

## Why?

Staying up-to-date.

## How?

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-5.html#breaking-changes
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-6.html#breaking-changes
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html#breaking-changes
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-8.html#correctness-fixes-and-breaking-changes
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#correctness-fixes-and-breaking-changes

## Testing Instructions

```sh
rm -rf packages/*/build-types packages/*/tsconfig.tsbuildinfo
npm run clean:package-types
npm run build:package-types
```

### CODE REVIEW

File: packages/components/src/icon/index.tsx [hunk 1]
```diff
@@ -44,9 +47,10 @@ type AdditionalProps< T > = T extends ComponentType< infer U >
 	? SVGProps< SVGSVGElement >
 	: {};
 
-export type Props< P > = BaseProps< P > & AdditionalProps< IconType< P > >;
+export type Props< P extends {} > = BaseProps< P > &
+	AdditionalProps< IconType< P > >;
 
-function Icon< P >( {
+function Icon< P extends {} >( {
```
Comment (kevin940726) [inline packages/components/src/icon/index.tsx]: These `<Icon>` type changes are related to this breaking change in 4.8: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-8.html#unconstrained-generics-no-longer-assignable-to-

I'm not entirely sure that this is the right fix though. I'll defer this question to the component squad to double-check. c.c. @mirka

File: packages/keycodes/src/index.js [hunk 2]
```diff
@@ -151,7 +151,7 @@ export { isAppleOS };
 /**
  * Map the values of an object with a specified callback and return the result object.
  *
- * @template T
+ * @template {{ [s: string]: any; } | ArrayLike<any>} T
```
Comment (kevin940726) [inline packages/keycodes/src/index.js:154]: Related to https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-8.html#unconstrained-generics-no-longer-assignable-to-

File: packages/dom/src/focusable.js [hunk 3]
```diff
@@ -87,7 +87,7 @@ function isValidFocusableArea( element ) {
  * Returns all focusable elements within a given context.
  *
  * @param {Element} context              Element in which to search.
- * @param {Object}  [options]
+ * @param {Object}  options
```
Comment (kevin940726) [inline packages/dom/src/focusable.js:90]: See https://github.com/microsoft/TypeScript/issues/49869.

File: packages/components/src/icon/index.tsx (hunk 1 above)
Comment (ciampo) [reply packages/components/src/icon/index.tsx]: would changing this to 

```suggestion
function Icon< P = {} >( {
```

remove the need for all of the other related changes in the components package?

Comment (mirka) [reply packages/components/src/icon/index.tsx]: I'm skeptical whether we even need the generic. `React.ComponentType` is `< P = {} >` by default, and none of our usages are actually passing in explicit props types. I can't really imagine a situation where someone would want/need to do so, either.

I'd be ok with removing all the generics here for now 👉 https://github.com/WordPress/gutenberg/commit/c70c1d6884a291376a00abaa3ebafaa820079148

Comment (kevin940726) [reply packages/components/src/icon/index.tsx]: Great suggestions! I removed the generic in [a28ad60](https://github.com/WordPress/gutenberg/pull/48299/commits/a28ad60d96e8b40d3cc9afa3303e0bf94632dd82).

**Review by mirka — APPROVED** [review]
Looks good, thank you! 🚀

**Pull request merged**